middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -478,24 +538,31 @@ and let_cont_not_inlined env res k handler body =
      expression. *)
   let wrap, env, res = Env.flush_delayed_lets ~mode:Branching_point env res in
   let is_exn_handler = Continuation_handler.is_exn_handler handler in
-  let vars, arity, handler, res = continuation_handler env res handler in
+  let vars, arity, handler, free_vars_of_handler, res =
+    continuation_handler env res handler
+  in
   let catch_id, env =
     Env.add_jump_cont env k ~param_types:(List.map snd vars)
   in
-  let cmm, res =
+  let cmm, free_vars_of_body, res =
     (* Exception continuations are translated specially -- these will be reached
        via the raising of exceptions, whereas other continuations are reached
        using a normal jump. *)
     if is_exn_handler
     then let_cont_exn_handler env res k body vars handler ~catch_id arity
     else
       let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
-      let body, res = expr env res body in
+      let body, free_vars, res = expr env res body in
       ( C.create_ccatch ~rec_flag:false ~body
           ~handlers:[C.handler ~dbg catch_id vars handler],
+        free_vars,
```

Comment:
I would prefer if we didn't return the free variables of the body along with the whole catch expression. Could you do the union here (and pass `free_names_of_handler` to `let_cont_exn_handler` so that it can do the same) ?

Reply:
done.